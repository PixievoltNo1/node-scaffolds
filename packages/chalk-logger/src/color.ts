import type { Chalk } from 'chalk'
import chalk from 'chalk'

export type IColor = string | [number, number, number]

const colorKeywords = new Set([
  'black',
  'red',
  'green',
  'yellow',
  'blue',
  'magenta',
  'cyan',
  'white',
  'gray',
])

const brightColorKeywords = new Set([
  'redBright',
  'greenBright',
  'yellowBright',
  'blueBright',
  'magentaBright',
  'cyanBright',
  'whiteBright',
])

export function color2chalk(color: IColor, fg: boolean): Chalk {
  if (typeof color === 'string') {
    if (colorKeywords.has(color) || brightColorKeywords.has(color)) {
      if (!fg) {
        // eslint-disable-next-line no-param-reassign
        color = 'bg' + color[0].toUpperCase() + color.slice(1)
      }
      return chalk[color].bind(chalk)
    }
    try {
      return fg ? chalk.keyword(color) : chalk.bgKeyword(color)
    } catch (error) {
      return fg ? chalk.hex(color) : chalk.bgHex(color)
    }
  }
  return fg ? chalk.rgb(color[0], color[1], color[2]) : chalk.bgRgb(color[0], color[1], color[2])
}

export class ColorfulChalk {
  public readonly fg: Chalk
  public readonly bg: Chalk | null

  constructor(fg: IColor, bg?: IColor) {
    this.fg = color2chalk(fg, true)
    if (bg == null) this.bg = null
    else this.bg = color2chalk(bg, false)
  }
}
