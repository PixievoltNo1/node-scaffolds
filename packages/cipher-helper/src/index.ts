export * from './_base'
export * from './aes'
export * from './catalog'
export * from './types/catalog'
export * from './types/cipher-helper'
export * from './util/buffer'
export * from './util/key'
