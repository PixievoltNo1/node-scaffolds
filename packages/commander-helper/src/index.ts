export * from './command'
export * from './fs'
export * from './git'
export * from './merge'
export * from './option'
export * from './yarn'
