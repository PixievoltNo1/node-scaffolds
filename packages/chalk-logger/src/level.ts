import { ColorfulChalk } from './color'

export enum Level {
  DEBUG = 'debug',
  VERBOSE = 'verbose',
  INFO = 'info',
  WARN = 'warn',
  ERROR = 'error',
  FATAL = 'fatal',
}

export const DEBUG = Level.DEBUG
export const VERBOSE = Level.VERBOSE
export const INFO = Level.INFO
export const WARN = Level.WARN
export const ERROR = Level.ERROR
export const FATAL = Level.FATAL

export const levelOrdinalMap: Record<Level, number> = {
  [Level.DEBUG]: 1,
  [Level.VERBOSE]: 2,
  [Level.INFO]: 3,
  [Level.WARN]: 4,
  [Level.ERROR]: 5,
  [Level.FATAL]: 6,
}

export interface ILevelStyle {
  title: string
  header: ColorfulChalk
  headerDelimiter: ColorfulChalk
  content: ColorfulChalk
}

export type ILevelStyleMap = Record<Level, ILevelStyle>

export const defaultLevelStyleMap: ILevelStyleMap = Object.freeze({
  [Level.DEBUG]: {
    title: 'debug',
    header: new ColorfulChalk('grey'),
    headerDelimiter: new ColorfulChalk('grey'),
    content: new ColorfulChalk('grey'),
  },
  [Level.VERBOSE]: {
    title: 'verb ',
    header: new ColorfulChalk('cyan'),
    headerDelimiter: new ColorfulChalk('grey'),
    content: new ColorfulChalk('cyan'),
  },
  [Level.INFO]: {
    title: 'info ',
    header: new ColorfulChalk('green'),
    headerDelimiter: new ColorfulChalk('grey'),
    content: new ColorfulChalk('green'),
  },
  [Level.WARN]: {
    title: 'warn ',
    header: new ColorfulChalk('yellow'),
    headerDelimiter: new ColorfulChalk('grey'),
    content: new ColorfulChalk('yellow'),
  },
  [Level.ERROR]: {
    title: 'error',
    header: new ColorfulChalk('red'),
    headerDelimiter: new ColorfulChalk('grey'),
    content: new ColorfulChalk('red'),
  },
  [Level.FATAL]: {
    title: 'fatal',
    header: new ColorfulChalk('black', 'red'),
    headerDelimiter: new ColorfulChalk('grey'),
    content: new ColorfulChalk('redBright'),
  },
})

export const resolveLevel = (level: string): Level | null => {
  switch (level.toLowerCase()) {
    case 'debug':
      return Level.DEBUG
    case 'verbo':
    case 'verbose':
      return Level.VERBOSE
    case 'info':
    case 'information':
      return Level.INFO
    case 'warn':
    case 'warning':
      return Level.WARN
    case 'error':
      return Level.ERROR
    case 'fatal':
      return Level.FATAL
    default:
      return null
  }
}
