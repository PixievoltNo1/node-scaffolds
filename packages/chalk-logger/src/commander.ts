import type { Mutable } from '@guanghechen/utility-types'
import { resolveLevel } from './level'
import type { ILoggerFlags, ILoggerOptions } from './logger'

/**
 * Commander options
 */
interface ICommanderOptions {
  /**
   * logger level
   */
  logLevel?: string
  /**
   * logger name
   */
  logName?: string
  /**
   * logger mode
   */
  logMode?: string
  /**
   * logger flags
   */
  logFlag?: string[]
  /**
   * filepath of log output
   */
  logFilepath?: string
  /**
   * encoding of log file
   */
  logEncoding?: BufferEncoding
}

interface ICommander {
  option(flags: string, description?: string, defaultValue?: string | boolean): this
  option(flags: string, description: string, regexp: RegExp, defaultValue?: string | boolean): this
  option<T>(
    flags: string,
    description: string,
    fn: (value: string, previous: T) => T,
    defaultValue?: T,
  ): this
}

/**
 * register to commander
 * @param program {commander.Command}
 */
export function registerCommanderOptions(program: ICommander): void {
  program
    .option('--log-level <level>', "specify logger's level.")
    .option('--log-name <name>', "specify logger's name.")
    .option("--log-mode <'normal' | 'loose'>", "specify logger's name.")
    .option(
      '--log-flag <option>',
      "specify logger' option. [[no-]<date|title|colorful|inline>]",
      (val: string, acc: string[]) => acc.concat(val),
      [],
    )
}

export function parseOptionsFromCommander(commanderOptions: ICommanderOptions): ILoggerOptions {
  const flags: Partial<Mutable<ILoggerFlags>> = {}
  const options: ILoggerOptions = { flags }

  // resolve log level
  if (commanderOptions.logLevel != null) {
    const logLevel = commanderOptions.logLevel.toLowerCase()
    const level = resolveLevel(logLevel)
    if (level != null) options.level = level
  }

  // resolve log name
  if (commanderOptions.logName != null) {
    options.name = commanderOptions.logName
  }

  // resolve log mode
  if (commanderOptions.logMode != null) {
    const mode = commanderOptions.logMode.toLowerCase()
    if (['normal', 'loose'].includes(mode)) {
      options.mode = mode as 'normal' | 'loose'
    }
  }

  // resolve log flags
  if (commanderOptions.logFlag != null) {
    for (let flag of commanderOptions.logFlag) {
      let negative = false
      if (/^no-/.test(flag)) {
        negative = true
        flag = flag.slice(3)
      }
      flag = flag.toLowerCase()
      switch (flag) {
        case 'inline':
          flags.inline = !negative
          break
        case 'date':
          flags.date = !negative
          break
        case 'title':
          flags.title = !negative
          break
        case 'colorful':
          flags.colorful = !negative
          break
      }
    }
  }
  return options
}

export function partOptionsFromArgs(args: string[]): ILoggerOptions {
  const options: ICommanderOptions = { logFlag: [] }
  const regex = /^--log-([\w]+)(?:=([-\w]+))?/
  for (let i = 0; i < args.length; ++i) {
    const arg = args[i]
    const match = regex.exec(arg)
    if (match == null) continue
    let [, key, val] = match

    if (val == null) {
      if (i + 1 < args.length) {
        const nextArg = args[i + 1]
        if (/^-/.test(nextArg)) continue
        i += 1
        val = nextArg
      }
    }

    key = key.toLowerCase()
    switch (key) {
      case 'encoding':
        options.logEncoding = val as BufferEncoding
        break
      case 'filepath':
        options.logFilepath = val
        break
      case 'flag':
        options.logFlag!.push(val)
        break
      case 'level':
        options.logLevel = val
        break
      case 'mode':
        options.logMode = val
        break
      case 'name':
        options.logName = val
        break
    }
  }

  return parseOptionsFromCommander(options)
}
