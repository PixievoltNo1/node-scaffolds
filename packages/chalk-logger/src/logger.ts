import { isString } from '@guanghechen/helper-is'
import type { ILogger, Mutable } from '@guanghechen/utility-types'
import dayjs from 'dayjs'
import { normalizeString } from './format'
import type { ILevelStyleMap } from './level'
import { Level, defaultLevelStyleMap, levelOrdinalMap } from './level'

export interface ILoggerFlights {
  readonly date: boolean
  readonly title: boolean
  readonly inline: boolean
  readonly colorful: boolean
}

export interface ILoggerOptions {
  name?: string
  mode?: 'normal' | 'loose'
  level?: Level | null
  levelStyleMap?: ILevelStyleMap
  flights?: Partial<ILoggerFlights>
  placeholderRegex?: RegExp
  write?(text: string): void
}

export class Logger implements ILogger {
  public readonly name: string
  public readonly mode: 'normal' | 'loose' = 'normal'
  public readonly level: Level
  public readonly levelStyleMap: ILevelStyleMap
  public readonly flights: ILoggerFlights
  public readonly placeholderRegex: RegExp = /(?<!\\)\{\}/g
  public readonly write: (text: string) => void

  constructor(options?: ILoggerOptions) {
    this.name = ''
    this.level = options?.level ?? Level.INFO
    this.levelStyleMap = options?.levelStyleMap ?? defaultLevelStyleMap
    this.write = text => {
      process.stdout.write(text)
    }
    this.flights = {
      date: false,
      title: true,
      inline: false,
      colorful: true,
    }
    this.init(options)
  }

  public init(options?: ILoggerOptions): void {
    if (!options) return

    const self = this as Mutable<this>

    // Set logger name.
    self.name = options.name ?? self.name

    // Set logger mode.
    self.mode = options.mode ?? self.mode

    // Set logger level.
    self.level = options.level ?? self.level
    self.levelStyleMap = options.levelStyleMap ?? self.levelStyleMap

    // Set logger flights.
    if (options.flights) {
      const selfFlights = self.flights as Mutable<ILoggerFlights>
      selfFlights.date = options.flights.date ?? selfFlights.date
      selfFlights.title = options.flights.title ?? selfFlights.title
      selfFlights.inline = options.flights.inline ?? selfFlights.inline
      selfFlights.colorful = options.flights.colorful ?? selfFlights.colorful
    }

    // set placeholderRegex
    if (options.placeholderRegex != null) {
      let flags: string = this.placeholderRegex.flags
      if (!flags.includes('g')) flags += 'g'
      self.placeholderRegex = new RegExp(options.placeholderRegex.source, `${flags}`)
    }

    // Set logger writer.
    self.write = options.write ?? self.write
  }

  // format a log record.
  public format(level: Level, header: string, message: string): string {
    const content: string = this.formatContent(level, message)
    return header.length > 0 ? header + ' ' + content : content
  }

  // format a log record's header.
  public formatHeader(level: Level, date: Date): string {
    const dateText: string = this.flights.date
      ? this.formatContent(level, dayjs(date).format('YYYY-MM-DD HH:mm:ss'))
      : ''

    const levelStyle = this.levelStyleMap[level]
    let levelText = levelStyle.title
    if (this.flights.colorful) {
      levelText = levelStyle.labelChalk.fg(levelText)
      if (levelStyle.labelChalk.bg != null) levelText = levelStyle.labelChalk.bg(levelText)
    }

    const titleText: string = this.flights.title
      ? this.formatContent(level, '[' + this.name + ']')
      : ''

    let result = ''
    if (dateText) result += dateText + ' '
    result += levelText
    if (titleText) result += ' ' + titleText
    return result
  }

  public formatContent(level: Level, message: string): string {
    let text: string = message
    if (this.flights.colorful) {
      const levelStyle = this.levelStyleMap[level]
      text = levelStyle.contentChalk.fg(text)
      if (levelStyle.contentChalk.bg != null) {
        text = levelStyle.contentChalk.bg(text)
      }
    }
    return text
  }

  // format a log record part message according its type.
  public formatSingleMessage(message: unknown): string {
    return normalizeString(message ? message : String(message), this.flights.inline)
  }

  public debug(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(Level.DEBUG, messageFormat, ...messages)
  }

  public verbose(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(Level.VERBOSE, messageFormat, ...messages)
  }

  public info(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(Level.INFO, messageFormat, ...messages)
  }

  public warn(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(Level.WARN, messageFormat, ...messages)
  }

  public error(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(Level.ERROR, messageFormat, ...messages)
  }

  public fatal(messageFormat: string | unknown, ...messages: unknown[]): void {
    this.log(Level.FATAL, messageFormat, ...messages)
  }

  // write a log record.
  public log(level: Level, messageFormat: string | unknown, ...messages: unknown[]): void {
    if (!level || levelOrdinalMap[level] < levelOrdinalMap[this.level]) return
    const header = this.formatHeader(level, new Date())

    let newline = false
    const formatPattern: string = isString(messageFormat) ? messageFormat : ''
    const items: string[] = (isString(messageFormat) ? messages : [messageFormat, ...messages]).map(
      msg => {
        const text = this.formatSingleMessage(msg)
        if (text.endsWith('\n')) {
          newline = true
          return '\n' + text
        }
        return text
      },
    )

    let unpairedIdx = 0
    let message =
      items.length > 0
        ? formatPattern.replace(this.placeholderRegex, m => {
            const value = items[unpairedIdx]
            unpairedIdx += 1
            return value === undefined ? m : value
          })
        : formatPattern
    if (unpairedIdx < items.length) message += ' ' + items.slice(unpairedIdx).join(' ')
    if (!newline && !message.endsWith('\n')) message += '\n'

    switch (this.mode) {
      case 'loose':
        this.write('\n' + this.format(level, header, message) + '\n')
        break
      case 'normal':
      default:
        this.write(this.format(level, header, message))
        break
    }
  }
}
