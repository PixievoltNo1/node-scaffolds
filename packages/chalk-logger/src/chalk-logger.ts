import type { Mutable } from '@guanghechen/types'
import { parseOptionsFromArgs } from './commander'
import type { ILoggerOptions } from './logger'
import { Logger } from './logger'

export class ChalkLogger extends Logger {
  /**
   * prefix of logger.name
   */
  protected basename: string | null = null
  /**
   * name passed into .setName()
   */
  protected divisionName: string | null = null

  constructor(options?: ILoggerOptions, args?: string[]) {
    const optionsFromArgs = parseOptionsFromArgs(args || [])
    const resolvedOptions = {
      ...options,
      ...optionsFromArgs,
      flights: { ...options?.flights, ...optionsFromArgs.flights },
    }

    super(resolvedOptions)
    const basename = resolvedOptions?.name ?? ''
    this.setBaseName(basename)
  }

  /**
   * update logger's name
   * @param name
   */
  public setName(name: string | null): void {
    const resolvedName: string = [this.basename, name]
      .filter((x): x is string => x != null && x.length > 0)
      .join(' ')
    const self = this as Mutable<this>
    self.name = resolvedName
    this.divisionName = name
  }

  /**
   * update basename of logger
   * @param basename
   */
  public setBaseName(basename: string | null): void {
    this.basename = basename
    this.setName(this.divisionName)
  }

  /**
   * update logger's mode
   * @param mode
   */
  public setMode(mode: 'normal' | 'loose'): void {
    const self = this as Mutable<this>
    self.mode = mode
  }
}
