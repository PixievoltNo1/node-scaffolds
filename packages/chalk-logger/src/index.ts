import type { Mutable } from '@guanghechen/utility-types'
import { partOptionsFromArgs } from './commander'
import type { Level } from './level'
import type { ILoggerOptions } from './logger'
import { Logger } from './logger'

export * from './color'
export * from './commander'
export * from './level'
export * from './logger'

export class ChalkLogger extends Logger {
  /**
   * prefix of logger.name
   */
  protected basename: string | null = null
  /**
   * name passed into .setName()
   */
  protected divisionName: string | null = null

  constructor(options?: ILoggerOptions, args?: string[]) {
    const optionsFromArgs = partOptionsFromArgs(args || [])
    const resolvedOptions = {
      ...options,
      ...optionsFromArgs,
      flags: { ...options?.flags, ...optionsFromArgs.flags },
    }

    super(resolvedOptions)
    const basename = resolvedOptions?.name ?? ''
    this.setBaseName(basename)
  }

  /**
   * update logger's level
   * @param level
   */
  public setLevel(level: Level | null | undefined): void {
    if (level == null) return
    const self = this as Mutable<this>
    self.level = level
  }

  /**
   * update logger's name
   * @param name
   */
  public setName(name: string | null): void {
    const resolvedName: string = [this.basename, name]
      .filter((x): x is string => x != null && x.length > 0)
      .join(' ')
    const self = this as Mutable<this>
    self.name = resolvedName
    this.divisionName = name
  }

  /**
   * update basename of logger
   * @param basename
   */
  public setBaseName(basename: string | null): void {
    this.basename = basename
    this.setName(this.divisionName)
  }

  /**
   * update logger's mode
   * @param mode
   */
  public setMode(mode: 'normal' | 'loose'): void {
    const self = this as Mutable<this>
    self.mode = mode
  }
}

export default ChalkLogger
