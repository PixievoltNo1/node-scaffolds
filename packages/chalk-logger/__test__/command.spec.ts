import { Level, parseOptionsFromArgs, registerCommanderOptions } from '../src'

describe('command funcs', function () {
  test('generateOptions', function () {
    expect(
      parseOptionsFromArgs([
        '--log-level=debug',
        '--log-name=waw',
        '--log-mode=loose',
        '--log-flag=date',
        '--log-flag=no-inline',
        '--log-flag=no-title',
        '--log-flag',
        'no-colorful',
      ]),
    ).toEqual({
      name: 'waw',
      mode: 'loose',
      level: Level.DEBUG,
      flags: {
        date: true,
        title: false,
        inline: false,
        colorful: false,
      },
    })
  })

  test('registerCommanderOptions', function () {
    const infos: any[] = []

    const program = {
      option: (...args: any[]) => {
        infos.push(args)
        return program
      },
    }

    registerCommanderOptions(program)
    expect(infos).toMatchSnapshot('registered options')
  })
})
