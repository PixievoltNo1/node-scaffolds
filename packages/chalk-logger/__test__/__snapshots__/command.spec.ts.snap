// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command funcs registerCommanderOptions: registered options 1`] = `
[
  [
    "--log-level <level>",
    "specify logger's level.",
  ],
  [
    "--log-name <name>",
    "specify logger's name.",
  ],
  [
    "--log-mode <'normal' | 'loose'>",
    "specify logger's name.",
  ],
  [
    "--log-flag <option>",
    "specify logger' option. [[no-]<date|title|colorful|inline>]",
    [Function],
    [],
  ],
]
`;
